src/Microsoft.AspNet.OData.Shared/Routing/ODataVersionConstraint.cs

```diff
@@ -0,0 +1,126 @@
+        /// </summary>
+        public ODataVersion Version { get; private set; }
+
+        /// <summary>
+        /// If set to true, allow passing in both OData V4 and previous version headers.
+        /// </summary>
+        public bool IsRelaxedMatch { get; set; }
+
+        /// <summary>
+        /// Determine if there is a version match.
+        /// </summary>
+        /// <param name="headers">The request headers.</param>
+        /// <param name="serviceVersion">The supported service version.</param>
+        /// <param name="maxServiceVersion">The max supported service version.</param>
+        /// <returns>True if there is a match; false otherwise.</returns>
+        private bool IsVersionMatch(IDictionary<string, IEnumerable<string>> headers, ODataVersion? serviceVersion, ODataVersion? maxServiceVersion)
+        {
+            // The match behavior depends on value of IsRelaxedMatch.
+            // If users select using relaxed match logic, the header contains both V3 (or before) and V4 style version
+            // will be regarded as valid. While under non-relaxed match logic, both version headers presented will be
+            // regarded as invalid. The behavior for other situations are the same. When non version headers present,
+            // assume using V4 version.
+            if (!ValidateVersionHeaders(headers))
+            {
+                return false;
+            }
+
+            ODataVersion? requestVersion = GetVersion(headers, serviceVersion, maxServiceVersion);
+            return requestVersion.HasValue && requestVersion.Value == Version;
+        }
+
+        private bool ValidateVersionHeaders(IDictionary<string, IEnumerable<string>> headers)
+        {
+            bool containPreviousVersionHeaders =
+                headers.ContainsKey(PreviousODataVersionHeaderName) ||
+                headers.ContainsKey(PreviousODataMinVersionHeaderName) ||
+                headers.ContainsKey(PreviousODataMaxVersionHeaderName);
+
+            bool containPreviousMaxVersionHeaderOnly =
+                headers.ContainsKey(PreviousODataMaxVersionHeaderName) &&
+                !headers.ContainsKey(PreviousODataVersionHeaderName) &&
+                !headers.ContainsKey(PreviousODataMinVersionHeaderName);
+
+            bool containCurrentMaxVersionHeader = headers.ContainsKey(ODataMaxServiceVersionHeader);
+
+            return IsRelaxedMatch
+                ? !containPreviousVersionHeaders || (containCurrentMaxVersionHeader && containPreviousMaxVersionHeaderOnly)
+                : !containPreviousVersionHeaders;
+        }
+
+        private ODataVersion? GetVersion(IDictionary<string, IEnumerable<string>> headers, ODataVersion? serviceVersion, ODataVersion? maxServiceVersion)
+        {
+            // The logic is as follows. We check OData-Version first and if not present we check OData-MaxVersion.
+            // If both OData-Version and OData-MaxVersion do not present, we assume the version is V4
```

Comment:
nit typo: If both OData-Version and OData-MaxVersion **are** not present

Reply:
Fixing...